### DISCUSSION TITLE
WebXR integrations and Lens distortion effect

### DISCUSSION THREAD

**Discussion by danrossi** [pmndrs/postprocessing#513] [Q&A]
I've integrated three webxr fixes for post processing, with some modifications in a fork and a custom lens distortion effect that seems to be working. However my use case for distortion is a webxr polyfill for Iphone, as the standard one is broken with three. 

I am requiring to distort each webxr camera, right now what I have is it shows half distortion for each eye because its applying to the whole viewport. 

Is there extra things required in my effect to apply for each eye ? 

postprocessing demos I've updated showing initial webxr integration

https://danrossi.github.io/postprocessing/public/demo/#bloom

Required fixes

https://github.com/danrossi/postprocessing/blob/main/src/core/EffectComposer.js#L240
https://github.com/danrossi/postprocessing/blob/main/src/passes/EffectPass.js#L579
https://github.com/danrossi/postprocessing/blob/main/src/passes/RenderPass.js#L291
https://github.com/danrossi/postprocessing/blob/main/src/passes/RenderPass.js#L328

The required three fixes

https://github.com/mrdoob/three.js/pull/26160

Reference

https://github.com/pmndrs/postprocessing/issues/452
https://github.com/pmndrs/postprocessing/issues/440

Lens distortion examples
https://danrossi.github.io/three-lens-distortion/example/postprocessing.html
https://danrossi.github.io/three-lens-distortion/example/postprocessing-video.html

The effect and shader
https://github.com/danrossi/three-lens-distortion/blob/main/src/LensDistortionEffect.js
https://github.com/danrossi/three-lens-distortion/blob/main/src/glsl/lens-distortion.frag

Each eye is a half distortion for fullscreen

![Screenshot (90)](https://github.com/pmndrs/postprocessing/assets/710277/4e040e9d-a078-438f-ae81-f8ee8dafab5d)

I'm trying to replicate it looking like this. 

![Screenshot (89)](https://github.com/pmndrs/postprocessing/assets/710277/539fac44-8553-4984-a3fa-5e2e776cae28)

Any thoughts would be appreciated.

**Comment by vanruesc** [comment]
> I'm trying to replicate it looking like this

https://github.com/pmndrs/postprocessing/issues/452#issuecomment-1480245175

In that thread it was also mentioned that we should wait for https://github.com/mrdoob/three.js/pull/26160 to get merged first.

**Comment by danrossi** [comment]
I have a branch with the integrated webxr fix by updating npm in that branch. Although I copied over a bundle with merged updated three. It seems to be working with the changes I made. But it becomes fullscreen effects. I have to manipulate the new effect to be applied on both half viewports. Im sure there is a trick,  maybe in a custom pass. 

I cant generate github pages of the manual to show the demo updates I made working. Screenshot here

https://github.com/pmndrs/postprocessing/issues/452#issuecomment-1635968230 

Right now each eye is half the fullscreen effect on each side. I have to lens distort each eye individually. 

https://github.com/danrossi/postprocessing/tree/webxr-render-fix

**Comment by danrossi** [comment]
I managed to deploy the manual demos here showing XR from that branch working but its on cloudfront so needs index.html on the ends of the urls. 

https://dev.electroteque.org/postprocessing/manual/demos/light-shadow/bloom/index.html

**Comment by vanruesc** [comment]
>  But it becomes fullscreen effects. I have to manipulate the new effect to be applied on both half viewports. Im sure there is a trick

See the last part of [this answer](https://github.com/pmndrs/postprocessing/pull/514#issuecomment-1636800332). Postprocessing doesn't support viewport/scissor settings yet.

> I managed to deploy the manual demos here showing XR from that branch working but its on cloudfront so needs index.html on the ends of the urls.

The manual has intentionally not been published yet. It's no big deal, but I'd appreciate it if you could unpublish that.

**Reply by danrossi** [reply]
Its just a test to show my XR fixes working against that PR. You want it down ? I'll make a single simplified test then. I'll add a PR with the fixes when its ready. I think you suggested something that might help me figure out something with scissor test.

**Comment by danrossi** [comment]
I've moved my examples here with the lens effect. The video one I have to try and get it scissor tested then to apply both sides. Before the xr fixes in my branch both cameras are being rendered to one eye so is fullscreen effect now.  

https://danrossi.github.io/webxr-polyfill/examples/postprocessing.html
https://danrossi.github.io/webxr-polyfill/examples/postprocessing-video.html
